lightningd/plugin_hook.c

```diff
@@ -183,3 +184,41 @@ REGISTER_SYNC_PLUGIN_HOOK(db_write, db_write_hook_callback,
                           void *,
                           db_write_hook_serialize,
                           const char **);
+
+/* `sensitive_command` hook */
+
+static void sensitive_command_hook_serialize(struct command *cmd,
+                                             struct json_stream *s)
+{
+	json_object_start(s, "sensitive_command");
+	json_add_string(s, "command", cmd->json_cmd->name);
+	json_object_end(s);
+}
```

Comment:
This is, IMHO, way too narrow.  Just create a command hook, and call it.  Have a parameter as to whether it's sensitive maybe (?), but you should definitely include the entire json command parameters.

It should also be able to *alter* the commands, IMHO. This lets a plugin do other things, like logging all commands, or supporting new units (e.g. fiat amounts) or whatever.

Reply:
A sufficiently-well-designed hook, in combination with #2796 , would allow the command to be replaced completely, i.e. the hook might allow: `{ "passthrough": true }` to mean no change in the command, `{ "replace-command": { "method" : "foo", "params":{ "arg1" : "arg1" } } }` to replace the command with another command (or just its params), or `{ "return" : { "result" : 42 } }` or `{ "return" : { "error" : { "code" : 42, "message" : "life, the universe, and everything", "data" : "Earth" }}}` to not execute the command but return the specified result/error instead.

Reply:
The option that @ZmnSCPxj proposes may go a bit too far, but I'd also advocate for the entire payload at least being passed to the hook in order for the plugin to make a more informed decision.

I'd also drop the arbitrary distinction between sensitive and non-sensitive and just call the hook for all JSON-RPC calls that way we don't have two separate control flows (we can add the hook call directly in the JSON-RPC dispatch logic) and the plugin can decide what is sensitive and what's not (with the nice side-effect that we don't have to come up with a distinction that is going to be wrong anyway :wink:).

Reply:
> The option that @ZmnSCPxj proposes may go a bit too far, but I'd also advocate for the entire payload at least being passed to the hook in order for the plugin to make a more informed decision.

What I did (not already pushed) is passing the method name and params to the plugin, what do you think ?

> I'd also drop the arbitrary distinction between sensitive and non-sensitive and just call the hook for all JSON-RPC calls that way we don't have two separate control flows (we can add the hook call directly in the JSON-RPC dispatch logic) and the plugin can decide what is sensitive and what's not (with the nice side-effect that we don't have to come up with a distinction that is going to be wrong anyway wink).

That's part of the change for the next push :-)